### ISSUE TITLE
Lift 1:1 "Nullable on Without-Relation-Scalar Side" limitation

### ISSUE THREAD

**Issue by jasonkuhrt** [prisma/nexus-prisma#34]
#### Perceived Problem

- https://www.prisma.io/docs/concepts/components/prisma-schema/relations#the-side-without-a-relation-scalar-must-be-optional

#### Ideas / Proposed Solution(s)

- Configuration option to generate a guaranteed relation, essentially cheating what the resolver can actually using Prisma Client? Is this a problem in theory but never practice or a real risk of runtime bugs?

#### Workarounds

...

**Comment by nikolasburk** [comment]
> with the Prisma 1:1 limitation here it means that it is unsafe for Nexus Prisma to assume that both sides of the 1:1 relation will always be non-null, right?

Yes, Prisma can never guarantee that both sides of a 1-1 relation are required. So, enabling this constraint via nexus-prisma could lead to runtime issues where Nexus thinks the second side of a relation is non-null but then gets `null` returned from Prisma Client and hence can't fulfill its GraphQL "contract".

**Comment by jasonkuhrt** [comment]
@nikolasburk Thanks that's clear! While it might not be guarantee-able at the database level the user's application logic might guarantee it and thus at the API level at least they might want the non-nullable to hold.

This leads me to think allowing nexus-prisma users to OPT-IN into the generated output being non-null for the projection is OK as long as it is clearly documented that by doing this they are going against what Prisma Client can guarantee and therefore must in their business logic ensure it.

I am not 100% sure what "ensure it" would mean/look like, if you have any further insight or links I could read that would be great! 

I think it means that the user needs to be careful about how they delete records. For example if the backend logic deletes a `User` from the database then it must make sure to cascade delete the associated `Profile` or otherwise they would indeed be a in a state with a `Profile.user` being `null`.

Likewise if a `Profile` record is being created directly for some reason it must be connected to or inline-create a new `User` to ensure that it doesn't end up in a `Profile.user` null situation.

Does my thinking here make sense or am I way off track?

**Comment by vladyslav-n** [comment]
Dear @jasonkuhrt 
Thanks a lot for your work on this project, you're an absolute legend!

I think the issue you've raised here is of huge importance in relation to db/api schemas relation. I think this discussion here will be super active as soon as the plugin reaches its first production-ready states and more people start using it instead of the old plugin.

Of course, a database cannot guarantee both sides of a 1-1 relation exist, you can always create a "null" situation on one of those sides as a result of data manipulations. But my opinion is that there must be an option that would allow a developer to make such relations _required_ in the generated graphql schema instead of being optional just as it is at prisma schema level. I think developer should have an option to take that risk of possible discrepancy and take care of it at other levels. You're completely right that very commonly such discrepancy might happen as a result of not ideally handled **delete** operations, and the developer should take care of that logic with more attention in that case, so that discrepancy does not happen. I think that in many cases "correctly" (according to the underlying business logic) set required relations at the API level is the right thing to do though it might not be 100% identical to the constraints at the db level.
